### PR TITLE
HIVE-26447: Vectorization: wrong results when filter on repeating map key orc table

### PR DESCRIPTION
**Pull request by yigress**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -58,7 +58,7 @@
     <module>testutils</module>
     <module>packaging</module>
     <module>standalone-metastore</module>
-    <module>upgrade-acid</module>
```
Comment (t3rmin4t0r) [inline pom.xml:61]: Remove the pom.xml from the PR

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java [hunk 2]
```diff
@@ -66,10 +66,13 @@ public int findScalarInMap(MapColumnVector mapColumnVector, int mapBatchIndex) {
     byte[][] keyVector = keyColVector.vector;
     int[] keyStart = keyColVector.start;
     int[] keyLength = keyColVector.length;
+    final boolean isRepeating = keyColVector.isRepeating;
     for (int i = 0; i < count; i++) {
       final int keyOffset = offset + i;
+      final int len = isRepeating? keyLength[i]: keyLength[keyOffset];
```
Comment (t3rmin4t0r) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java:72]: Is that supposed to be [i] or [0]

The two rows look a bit wrong there.

I would write keyOffset = isRepeating ? 0 : offset + i

if the intention was to use =0 for all repeating.

Comment (yigress) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java:72]: I found that the key is set isRepeating only when there is a single key (count=1), but I am not sure if different orc versions may have multiple keys, if there are multiple keys in repeating, keyLength may have multiple entries.  Would it be better just deal with the isRepeating and count=1 separately?

Comment (prasanthj) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java:72]: Agree with Gopal. If isRepeating=true, then keyLength and keyVector will have to be looked at index = 0. I was also expecting something like below
```
final int len = isRepeating? keyLength[0]: keyLength[keyOffset];
```

Comment (yigress) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java]: Thank you for the review @prasanthj ! updated.  I only see isRepeating is true for one single key, but not sure if there can be multiple keys and still have isRepeating=true

Comment (prasanthj) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java]: Could you update the test case with multiple keys and see if that produces expected result?

**Comment by yigress** [comment]
updated with test cases of repeating with multiple keys and mixed keys.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java (hunk 2 above)
Comment (yigress) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java]: added test cases of repeating multiple keys and mixed keys.

Comment (yigress) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java]: @prasanthj appreciate your review!

**Comment by prasanthj** [comment]
+1

**Review by sunchao — APPROVED** [review]

**Comment by sunchao** [comment]
Test failures unrelated. Merging now.

**Comment by jfsii** [comment]
Hi @sunchao / @yigress - I think some of the test failures could be related to this patch.

http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3492/3/pipeline/439 (which is from the run on this PR)

[2022-08-05T04:08:23.465Z] [ERROR] org.apache.hadoop.hive.cli.split16.TestMiniLlapLocalCliDriver.testCliDriver[orc_map_key_repeating]  Time elapsed: 3.725 s  <<< FAILURE!
[2022-08-05T04:08:23.465Z] java.lang.AssertionError: 
[2022-08-05T04:08:23.465Z] Client Execution succeeded but contained differences (error code = 2) after executing orc_map_key_repeating.q 
[2022-08-05T04:08:23.465Z] See ./ql/target/tmp/log/hive.log or ./itests/qtest/target/tmp/log/hive.log, or check ./ql/target/surefire-reports or ./itests/qtest/target/surefire-reports/ for specific test cases logs.
[2022-08-05T04:08:23.465Z] 	at org.junit.Assert.fail(Assert.java:89)

This PR directly added orc_map_key_repeating.

**Comment by jfsii** [comment]
I've attempted to fix the situation - https://github.com/apache/hive/pull/3509

**Comment by sunchao** [comment]
@jfsii oops sorry, I'll take a look at the fix

**Comment by yigress** [comment]
sorry about this. thank you @sunchao @jfsii !

**Comment by jfsii** [comment]
> sorry about this. thank you @sunchao @jfsii !

No problem @yigress - it is an easy mistake to make. Thanks for your contribution to the project!

File: ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java [hunk 3]
```diff
@@ -66,10 +66,13 @@ public int findScalarInMap(MapColumnVector mapColumnVector, int mapBatchIndex) {
...
     int[] keyStart = keyColVector.start;
     int[] keyLength = keyColVector.length;
+    final boolean isRepeating = keyColVector.isRepeating;
     for (int i = 0; i < count; i++) {
       final int keyOffset = offset + i;
+      final int len = isRepeating? keyLength[0]: keyLength[keyOffset];
+      byte[] rowKey = isRepeating? keyVector[0]: keyVector[keyOffset];
       if (StringExpr.equal(key, 0, key.length,
-          keyVector[keyOffset], keyStart[keyOffset], keyLength[keyOffset])) {
+          rowKey, keyStart[keyOffset], len)) {
```
Comment (maswin) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/vector/expressions/VectorUDFMapIndexStringScalar.java:75]: keyStart[keyOffset] should also be assigned based on isRepeating key.

**Pull request merged**